src/main/java/com/devotedmc/ExilePearl/listener/WorldBorderListener.java

```diff
@@ -36,4 +38,23 @@ public void onPearlDecay(PearlDecayEvent e) {
 			});
 		}
 	}
+
+	@EventHandler(priority = EventPriority.NORMAL)
+	public void onInventoryClick(InventoryClickEvent e) {
+		boolean autoFree = pearlApi.getPearlConfig().getShouldAutoFreeWorldBorder();
+		if (!autoFree) {
+			return;
+		}
+		boolean clickedTop = e.getRawSlot() < e.getView().getTopInventory().getSize();
```

Comment:
This won't include shift clicking or number keys. To avoid complex custom logic, could you just check whether the item clicked/on the cursor is a pearl (maybe differentiate between those two with a switch/case) and if it is and the open container is outside worldborder, always tell them something like "This chest is outside of world border, pearls left in it will automatically be freed"